sdk/keyvault/Shared/ChallengeBasedAuthenticationPolicy.cs

```diff
@@ -0,0 +1,276 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for
+// license information.
+
+using Azure.Core;
+using Azure.Core.Http;
+using Azure.Core.Pipeline;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Azure.Security.KeyVault
+{
+    internal class ChallengeBasedAuthenticationPolicy : HttpPipelinePolicy
+    {
+        private const string BearerChallengePrefix = "Bearer ";
+        private const string NoChallengeErrorMessage = "Failed to read WWW-Authenticate header from response";
+        private readonly TokenCredential _credential;
+
+        private AuthenticationChallenge _challenge = null;
+
+        private string _headerValue;
+
+        private DateTimeOffset _refreshOn;
+
+        public ChallengeBasedAuthenticationPolicy(TokenCredential credential)
+        {
+            _credential = credential;
+        }
+
+        public override void Process(HttpPipelineMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {
+            ProcessCoreAsync(message, pipeline, false).GetAwaiter().GetResult();
+        }
+
+        public override Task ProcessAsync(HttpPipelineMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {
+            return ProcessCoreAsync(message, pipeline, true);
+        }
+
+        private async Task ProcessCoreAsync(HttpPipelineMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
+        {
+            HttpPipelineRequestContent originalContent = message.Request.Content;
+
+            // if this policy doesn't have _challenge cached try to get it from the static challenge cache
+            _challenge ??= AuthenticationChallenge.GetChallenge(message);
+
+            // if we still don't have the challenge for the endpoint
+            // remove the content from the request and send without authentication to get the challenge
+            if (_challenge == null)
+            {
+                message.Request.Content = null;
+            }
+            // otherwise if we already know the challenge authenticate the request
+            else
+            {
+                await AuthenticateRequestAsync(message, async).ConfigureAwait(false);
+            }
+
+            if (async)
+            {
+                await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
+            }
+            else
+            {
+                ProcessNext(message, pipeline);
+            }
+
+            // if we get a 401
+            if (message.Response.Status == 401)
+            {
+                // set the content to the original content in case it was cleared
+                message.Request.Content = originalContent;
+
+                // update the cached challenge
+                var challenge = AuthenticationChallenge.GetChallenge(message);
+
+                if (_challenge == null || challenge != _challenge)
```

Comment:
`challenge != _challenge` is a reference comparison, is it intended?

Reply:
What is the intended behavior if there was no _challenge and none is received? Flowing the 401 up to the user I suspect?